Documentation/storage.md

```diff
@@ -0,0 +1,161 @@
+
+The following is an example of a custom `OAuth2Client` resource:
+
+```
+# Standard Kubernetes resource fields
+kind: OAuth2Client
+apiVersion: oidc.coreos.com/v1
+metadata:
+  namespace: foobar
+  name: client-foo
+
+# Custom fields defined by dex.
+clientID: "aclientid"
+clientSecret: "clientsecret"
+redirectURIs:
+- "https://app.example.com/callback"
+```
+
+The `ThirdPartyResource` type and the custom resources can be queried, deleted, and edited like any other resource using `kubectl`.
+
+```
+kubectl get thirdpartyresources # list third party resources registered on the clusters
+kubectl get --namespace=foobar oauth2clients # list oauth2 clients in a given namespace
+```
+
+To reduce administrative overhead, dex creates and manages its own third party resources and may create new ones during upgrades. While not strictly required we feel this is important for reasonable updates. Though, as a result, dex requires access to the non-namespaced `ThirdPartyResource` type. For example, clusters using RBAC authorization would need to create the following roles and bindings:
+
+```
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1alpha1
+metadata:
+  name: dex
+rules:
+  - apiGroups: ["oidc.coreos.com"] # API group created by dex
+    resources: ["*"]
+    verbs: ["*"]
+    nonResourceURLs: []
+  - apiGroups: ["extensions"]
+    resources: ["thirdpartyresources"]
+    verbs: ["create"] # To manage its own resources identity must be able to create thirdpartyresources.
+    nonResourceURLs: []
+---
+kind: ClusterRoleBinding
+apiVersion: rbac.authorization.k8s.io/v1alpha1
+metadata:
+  name: dex
+subjects:
+  - kind: ServiceAccount
+    name: dex                 # Service account assigned to the dex pod.
+    namespace: demo-namespace # The namespace dex is running in.
+roleRef:
+  kind: ClusterRole
+  name: identity
+  apiVersion: rbac.authorization.k8s.io/v1alpha1
+```
+
+The storage configuration is extremely limited since installations running outside a Kubernetes cluster would likely prefer a different storage option. An example configuration for dex running inside Kubernetes:
+
+```
+storage:
+  type: kubernetes
+  config:
+    inCluster: true
+```
+
+Dex determines the namespace it's running in by parsing the service account token automatically mounted into its pod.
+
+## SQL
+
+Dex supports two flavors of SQL, SQLite3 and Postgres. MySQL and CockroachDB may be added at a later time.
+
+Migrations are performed automatically on the first connection to the SQL server (it does not support rolling back). Because of this, dex requires being able to add and alter the tables for its database.
+
+NOTE: Previous versions of dex required symmetric keys to encrypt certain values before sending them to the database. This feature has not been ported to V2 yet, though it may be added later (but possibly without a migration path for users of the current implementation).
```

Comment:
I would reword the last second sentence to be

This feature has not yet been ported to dex v2.  If it is added later there may not be a migration path for current v2 users.